### DISCUSSION TITLE
Need to know about the feature extraction procedure from YOLO_v8 backbone

### DISCUSSION THREAD

**Discussion by 97mallikan** [ultralytics/ultralytics#5906] [General]
I am using YOLO_v8 to detect objects from video files. Now I want to perform feature extraction on frames of video files using the backbone section of that YOLO_v8 model.
Can anyone guide me on how can I do that? If there is any documentation or example that exists please share it with me, it would be very helpful to me.

Thanks

**Comment by pderrenger** [comment]
@97mallikan certainly! To perform feature extraction using the backbone of a YOLOv8 model, you can utilize the `embed` functionality provided by the YOLOv8 Python interface. This allows you to extract feature vectors from specified layers of the model.

Here's a concise example of how you can achieve this:

```python
from ultralytics import YOLO

# Load your YOLOv8 model
model = YOLO('yolov8n.pt')  # Replace with your model path if necessary

# Perform feature extraction on a video frame
# Specify the layers from which you want to extract features
layer_indices = [10, 14, 17]  # Example layer indices, adjust as needed
results = model.predict(source='path/to/your/video.mp4', embed=layer_indices)

# Iterate over results to access the embeddings
for result in results:
    embeddings = result.embeddings  # List of feature tensors from the specified layers
    # Now you can use these embeddings for your application
```

The `embed` parameter takes a list of layer indices from which you want to extract features. The `predict` method will return results that include the embeddings from these layers.

For more detailed information on using the `embed` parameter and other functionalities of YOLOv8, please refer to the Predict section in the Ultralytics Docs. This will provide you with additional insights and examples on how to work with the model for various tasks, including feature extraction.

**Reply by 97mallikan** [reply]
Thank you for your response.

**Reply by pderrenger** [reply]
@97mallikan absolutely, happy to help! 🚀 To extract features using the YOLOv8 model's backbone, you can use the `.predict()` method along with the `embed` argument. This allows you to specify which layers' outputs you'd like to grab. Here's a quick example:

```python
from ultralytics import YOLO

# Load model
model = YOLO('yolov8n.pt')  # Make sure to use your model path

# Perform prediction and extract features
results = model.predict('path/to/image.jpg', embed=[10, 14, 17])  # Adjust layers as needed

# Access the embeddings
features = results.embeddings
```

This snippet assumes you're interested in features from layers 10, 14, and 17. Just change those layer indices based on your requirements. Hope this clears things up! Let me know if there's anything else you're curious about. 😊

**Reply by AI-Elmanawy** [reply]
@pderrenger Thanks for your response but I get this error "AttributeError: 'Tensor' object has no attribute 'embeddings'" any suggestions please

**Reply by daisy9542** [reply]
> @pderrenger Thanks for your response but I get this error "AttributeError: 'Tensor' object has no attribute 'embeddings'" any suggestions please

You can check out this issue: https://github.com/ultralytics/ultralytics/issues/15678. Using the method of registering forward hooks is useful for me.

**Comment by 1999-ali** [comment]
Hello

Best wishes for 2025!

 I want to use the Yolo model, either Yolov8 or Yolov11, which I have not decided on yet. How can I extract the future and use it as a flattened layer with RNN models ?

**Reply by glenn-jocher** [reply]
@1999-ali thank you for your query! To extract features from YOLOv8 or YOLO11, you can forward your input through the model's backbone and retrieve the feature maps. These can then be flattened and used as input for your RNN. For a guide on extracting features, you can refer to this [blog post](https://www.ultralytics.com/blog/extracting-outputs-from-ultralytics-yolov8). Let me know if you need further details!

**Reply by 1999-ali** [reply]
Thank you very much for your response. I’m still struggling with the performance of my training using YOLO + LSTM. While other researchers use CNNs followed by LSTMs (since CNNs can process both images and sequential data), the idea is to improve performance by combining image information with sequential data and feeding it into the LSTM. If I’m not mistaken, YOLO is essentially a CNN-based model.

Does the YOLO model accept non-image inputs? Specifically, is it possible to input both images and sequential data into YOLO, and then pass the combined information to an LSTM?

**Reply by glenn-jocher** [reply]
Thank you for your question. YOLO models are designed to process image data and do not natively support non-image inputs like sequential data. However, you can extract features from YOLO's backbone (e.g., using the `embed` method) and combine these embeddings with your sequential data externally. The combined features can then be passed to an LSTM. This approach leverages YOLO's strength in image feature extraction while integrating temporal data through the LSTM.

**Comment by MjdMahasneh** [comment]
Hi! 

This solution doesn't seem to work anymore! I am working with YOLO11:

```python
from ultralytics import YOLO

# Load model
model = YOLO('yolo11n.pt')  # Make sure to use your model path

# Perform prediction and extract features
results = model.predict('./sample/1.png', embed=[10, 14, 17])  # Adjust layers as needed

print(f"Results: {results}")

# Access the embeddings
features = results.embeddings
```

and here is my output:

```
Results: [tensor([-1.5129e-01,  6.2808e-02,  1.0787e-01,  1.7923e-01, -2.0137e-01, -2.1150e-01, -7.8448e-02, -2.2041e-01, -1.7672e-01,  1.7080e-02, -1.6426e-01, -2.2463e-01, -1.9471e-01, -1.9638e-01,  1.5269e+00, -2.3549e-01,  3.8536e-01, -1.3250e-01, -1.8954e-01, -1.8129e-01, -2.2158e-01, -2.5010e-01, -1.9971e-01,  1.3551e-01,
        -2.2045e-01, -9.5793e-02, -2.1925e-01, -4.3828e-02, -1.3474e-01, -2.2052e-01, -8.2979e-02, -2.2576e-01, -1.2791e-01,  2.1180e-01,  4.0499e-01, -2.3374e-01, -6.5858e-02, -1.7846e-01,  1.6791e-01,  1.8391e-01, -2.3348e-01, -2.0792e-01, -1.9638e-01, -1.7818e-01, -1.4032e-01, -2.1874e-01, -4.0534e-02, -2.0114e-01,
        -1.6627e-01, -6.7030e-02, -2.2173e-01, -1.9873e-01, -2.0159e-01, -3.1278e-02, -2.2414e-01, -1.7254e-01, -1.7450e-01,  8.8256e-02, -2.0617e-01, -2.0037e-01, -1.8147e-01, -1.9994e-02, -2.2993e-01, -2.2939e-01, -2.2250e-01, -7.4227e-02, -1.2370e-01,  1.6250e-02, -2.1093e-01, -2.4699e-01, -2.2203e-01, -2.1168e-01,
        -1.6478e-01, -2.2690e-01, -4.8238e-02, -1.7088e-01, -1.3800e-01, -6.0106e-02, -1.4849e-01, -2.5381e-01, -5.7328e-02, -1.5100e-01, -1.9843e-01, -1.8389e-01,  2.1825e-01, -2.4097e-01, -2.2214e-01, -1.5662e-01, -1.8538e-01, -1.8577e-01, -1.4002e-01, -2.4194e-01, -1.9334e-01, -1.7060e-01, -2.2368e-01,  8.0441e-01,
         2.2000e-01, -2.0234e-01,  2.4399e-02, -1.6429e-01, -1.6829e-01,  2.2732e-01, -1.9536e-01,  5.2277e-01, -2.3435e-01, -2.2484e-01, -9.4745e-02, -2.3636e-01, -1.2562e-02, -2.0470e-01, -6.6598e-02, -1.9069e-01, -2.1749e-01, -1.6502e-01, -1.0159e-01,  6.8565e-02, -1.7222e-01, -1.7157e-01, -1.6568e-01,  6.4046e-02,
        -1.0239e-01, -5.9985e-02, -2.1751e-01, -1.6222e-01, -2.2257e-01,  7.5186e-02, -1.5780e-01, -2.4073e-01, -1.2521e-02, -1.7494e-01, -1.8742e-01, -1.1533e-01, -1.8278e-01,  1.2352e-01, -1.8346e-01,  6.5711e-02, -2.3270e-01, -8.8406e-02, -1.1072e-01,  1.9754e-02, -3.6533e-02, -2.2730e-01, -2.4399e-01, -1.3625e-01,
         2.9421e-01, -3.6468e-02, -2.1456e-01, -2.3514e-01,  3.0781e-01, -1.5491e-01, -1.9404e-01, -2.4370e-01,  7.7945e-03, -1.3748e-01, -2.2275e-01, -2.1927e-01, -9.7487e-02, -2.1106e-01, -1.6006e-01, -1.6906e-01, -1.5030e-01, -1.1474e-01, -2.1856e-01, -1.9524e-01, -2.2768e-01, -2.2838e-01, -4.4394e-02, -5.1301e-02,
        -1.5123e-01,  2.4326e-01, -1.9019e-01, -2.2153e-01, -6.9455e-03, -2.1624e-01, -2.1222e-01, -1.9984e-01, -2.0559e-01, -5.9719e-02, -2.0475e-01, -2.4775e-01, -1.8483e-01, -1.7309e-01, -1.8851e-01, -1.1770e-01, -1.6029e-01, -9.2364e-02, -1.4821e-02, -1.8247e-01, -2.1358e-01, -2.9332e-02, -1.6938e-01, -8.6703e-02,
        -1.9697e-01, -1.8581e-01, -2.2456e-01, -1.9733e-01, -1.8789e-01,  1.2145e-01,  4.0248e-02, -2.0672e-01,  3.2648e-01, -1.6117e-01, -1.8260e-01, -2.1857e-01, -2.1181e-01, -1.5422e-01, -2.8535e-02,  2.2093e-01, -2.0521e-01,  6.1841e-02,  1.2873e-01,  3.6051e-01, -1.5676e-01, -2.1419e-01, -1.5411e-01, -2.0407e-01,
        -9.9040e-02, -2.0740e-01, -2.2645e-01, -2.1574e-01, -1.3583e-01, -1.7420e-01,  8.4974e-02,  1.4242e-01, -1.3785e-01, -4.5225e-03, -1.4463e-01, -2.1389e-02, -2.4812e-01, -2.1542e-01, -1.7764e-01, -1.9014e-01, -1.0230e-01, -2.1366e-01, -1.5064e-01, -1.4707e-01, -2.3938e-01, -8.4202e-02, -3.0230e-02, -5.0107e-02,
         7.7778e-02,  4.2929e-01, -8.6467e-02, -2.3033e-01, -9.8744e-02, -3.3163e-02, -2.1290e-01,  2.2949e-01, -1.7050e-01, -1.8956e-01,  5.5552e-02, -2.2014e-01, -2.2391e-01, -1.8962e-01, -1.0653e-01,  9.7761e-02, -3.8180e-02, -4.3615e-03, -1.9857e-01, -3.7709e-02, -2.4137e-01, -1.8279e-01,  6.6304e-01, -2.4246e-01,
        -1.7658e-01, -2.1716e-01, -1.8653e-01, -1.9798e-01,  5.5049e-03,  2.1577e-01, -1.4279e-01,  1.7262e-01, -2.5943e-01,  7.8644e-02,  1.5758e-01,  1.5911e-01, -2.1716e-01, -1.9132e-01,  4.2883e-02, -1.4199e-01, -4.3046e-02, -1.6951e-01,  4.7185e-02, -1.8815e-01, -1.7751e-01,  1.2270e-02,  1.8017e-01,  3.5772e-01,
         2.3238e-02,  5.8928e-01, -2.1052e-01, -1.6133e-01,  1.6469e-01, -1.6600e-01, -2.0406e-01, -7.0860e-02, -6.6729e-02,  1.6641e-02, -2.0122e-01, -3.0579e-02, -1.7326e-01, -2.3784e-01,  2.1908e-01,  2.7699e-01,  1.4189e-01,  1.4023e-01, -2.2511e-01, -2.4236e-01, -2.1484e-01, -1.8634e-01, -1.4877e-01, -2.2019e-01,
         6.2560e-02, -2.2803e-01,  2.0718e-01,  1.1440e-02,  4.7638e-03,  9.0600e-01, -2.2256e-01,  3.6933e-03, -1.3236e-01, -1.7438e-01, -7.9871e-02, -2.0184e-01, -2.8899e-03, -3.0621e-02, -2.1016e-02, -2.3750e-01, -5.1431e-02, -1.1772e-01,  2.9843e-01, -2.3530e-01, -8.2855e-02, -9.2423e-02,  2.9610e-01,  5.7236e-01,
        -2.1516e-01, -2.6645e-01,  4.2225e-02,  1.4058e-01, -1.3641e-01, -1.8493e-02,  5.1577e-02,  5.1082e-01, -1.6720e-01, -2.4609e-01, -1.6685e-01,  5.0352e-01,  1.3596e-03, -6.3644e-02,  1.0461e-01, -7.2265e-02, -2.1088e-01,  4.0721e-01, -5.7004e-02,  2.7941e-01, -1.8644e-01,  1.3266e-01, -1.2623e-01, -1.7342e-03,
         2.6826e-01, -1.5446e-01,  5.5899e-01, -1.2144e-01, -2.1696e-01, -5.2888e-02, -1.0396e-01,  2.3186e-01, -2.8774e-02,  2.0974e-02, -1.5565e-01, -1.0304e-01, -1.9481e-01,  3.8612e-01,  5.5468e-02,  1.5246e-02,  3.9538e-02,  1.9967e-01, -3.5871e-03, -2.0874e-01,  1.2712e-01, -2.1911e-01,  7.0616e-02, -4.8327e-02,
        -2.1464e-01, -2.0907e-01, -1.9200e-01, -1.6533e-01, -2.0106e-01, -2.0006e-01, -5.9382e-02, -2.3331e-01, -9.2937e-02, -9.3496e-02, -1.4302e-01,  2.4630e-02,  1.5880e-01, -2.2141e-01, -1.2390e-01,  7.3635e-02, -1.6021e-01,  6.8235e-01, -1.9709e-01, -1.7501e-01,  1.9282e-01, -2.2361e-01,  1.6009e-02, -1.6242e-01,
        -1.5729e-01, -2.2807e-01, -2.3339e-01,  2.0141e-01, -2.2747e-01, -2.1929e-01, -7.4937e-02, -1.2813e-01, -1.3121e-01, -9.4018e-02,  1.2616e-01, -2.1220e-01,  2.9696e-01, -5.5680e-02, -1.5414e-01,  2.2512e-01,  4.0767e-01, -2.2008e-01, -1.9555e-01, -1.8141e-01, -1.7600e-01, -1.0458e-01,  8.1191e-02,  7.8994e-02,
         5.3812e-01, -2.1054e-01,  9.8393e-03, -1.9740e-01, -6.9733e-02, -7.8251e-02, -2.0219e-01, -2.5654e-01,  7.4605e-01, -2.4652e-01, -2.2247e-01, -1.3429e-01,  2.3164e-02, -2.6225e-01, -2.4926e-01, -1.7073e-01], device='cuda:0')]
Traceback (most recent call last):
  File "C:\Users\mma\Desktop\temp\scratch.py", line 12, in <module>
    features = results.embeddings
AttributeError: 'list' object has no attribute 'embeddings'
```



any help on this would be very much appreciated! I have tried hooks too with no luck!

**Reply by glenn-jocher** [reply]
To extract backbone features from YOLO11, use the `model.embed()` method instead of `predict()`. Here's a working example based on the latest implementation:

```python
from ultralytics import YOLO

model = YOLO('yolo11n.pt')
embeddings = model.embed('path/to/image.jpg')  # Returns list of tensors
print(embeddings[0].shape)  # Access first image's features
```

For more details, see the [model.embed documentation](https://docs.ultralytics.com/reference/engine/model/#ultralytics.engine.model.Model.embed) which shows the updated interface. The `embeddings` attribute has been replaced with direct return of feature tensors.

**Reply by MjdMahasneh** [reply]
@glenn-jocher thanks for following up. This snippet outputs the following:


```
[tensor([-0.0934,  0.0036, -0.1939, -0.2022,  0.0134, -0.0509,  0.2139,  0.0318,  0.2978, -0.1542, -0.0142,  0.0733, -0.2204, -0.2032,  0.2377, -0.2263,  0.4047, -0.0665,  0.0310, -0.1567, -0.2031,  0.0311, -0.1834, -0.2218, -0.1233, -0.2375, -0.1577,  0.2864, -0.2052,  0.0834, -0.2120,  0.0115,  0.2878, -0.1640,
        -0.2415,  0.2563, -0.2312, -0.0039, -0.1392, -0.1066,  0.1264,  0.5295,  0.2495, -0.2421,  0.3048, -0.1546, -0.0188, -0.1093, -0.2383,  0.0854,  0.0772,  0.2265,  0.5462,  0.1231,  0.0639,  0.1361,  0.5342, -0.1439,  0.1538, -0.2105, -0.2316,  0.3332, -0.0880, -0.2322, -0.0720, -0.0630,  0.3051, -0.2321,
         0.5776, -0.1364,  0.4692, -0.1576,  0.2371, -0.2319,  0.0547, -0.0634,  0.1623, -0.1950, -0.0215, -0.0585,  0.4984,  0.3661,  0.1071,  0.1284, -0.0614, -0.0329,  0.0259,  0.1524, -0.2229, -0.1796, -0.2075, -0.1829,  0.1590, -0.1665,  0.2487,  0.2630,  0.0854, -0.1857, -0.2454,  0.0619,  0.1228, -0.0605,
        -0.0862,  0.4245, -0.1028, -0.2426, -0.2345,  0.3818, -0.0403,  0.1972,  0.0889,  0.0573,  0.1698,  0.1646,  0.1437, -0.0083,  0.2630, -0.0489, -0.2508,  0.1288, -0.0598,  0.2442, -0.1448,  0.2315,  0.5739, -0.1521,  0.3510, -0.2056,  0.0684,  0.0525, -0.1746,  0.0088, -0.1071, -0.1539,  0.0836, -0.2152,
         0.1369, -0.2273, -0.0694, -0.2352, -0.0193, -0.2117, -0.2364,  0.1921, -0.2285, -0.1663, -0.1761, -0.2455,  0.5388, -0.0527, -0.1293, -0.2190, -0.2566,  0.4222,  0.0222, -0.2612,  0.0515,  0.2710,  0.0170,  0.1660,  0.8660, -0.1136,  0.3106, -0.2159, -0.0535,  0.0447,  0.0776, -0.0816,  0.2783, -0.2496,
        -0.2589,  0.1760, -0.0629, -0.0115, -0.0124,  0.1007, -0.1162, -0.1789, -0.0366, -0.1072, -0.2422,  0.0070, -0.2140, -0.1410, -0.0185, -0.1671,  0.0957, -0.0497, -0.1351,  0.1554, -0.2366, -0.2500,  0.0899,  0.1803,  0.0127, -0.0463, -0.0014,  0.0594,  0.1346, -0.1571, -0.2017, -0.2474,  0.3424, -0.0931,
        -0.0152, -0.2089,  0.0103, -0.1263, -0.0893, -0.1728, -0.2271,  0.0046,  0.5551,  0.3392, -0.2345, -0.0579,  0.2570, -0.1957, -0.2003, -0.1181, -0.2106, -0.2124,  0.2880, -0.2114,  0.1881, -0.1491,  0.0119, -0.0421, -0.1827,  0.2892, -0.0050, -0.0772,  0.0387,  0.1164,  0.3490, -0.1574,  0.0773,  0.1571,
        -0.1953,  0.2428,  0.3277, -0.2383, -0.1965,  0.0193, -0.0027, -0.1101,  0.1921,  0.0640, -0.2419,  0.2977, -0.2293, -0.2271,  0.2830,  0.0345, -0.1929, -0.1434], device='cuda:0')]

torch.Size([256])

```

Could you kindly help me understand the output? I expect to access the response of the backbone network (e.g., conv responses for all layers) but all am seeing here is a 256 long vector! My aim is to recycle these feature in a downstream task.

**Reply by glenn-jocher** [reply]
The 256-length vector you're seeing is the default flattened feature embedding from the penultimate layer. To access backbone conv layer outputs specifically, you'll need to specify layer indices in the `embed` parameter. For example:

```python
results = model.predict('image.jpg', embed=[4, 6, 9])  # backbone layer indices
features = [f.shape for f in results.embeddings]
```

The exact layer indices depend on your specific YOLO11 architecture variant. You can inspect the model architecture with `model.info()` to identify which layers correspond to backbone outputs. For more detailed guidance on layer selection, see the Model Inspection section in our documentation.

**Reply by MjdMahasneh** [reply]
Thanks @glenn-jocher 

Hi, this unfortunately doesn't solve it! here is the run:

```
Traceback (most recent call last):
  File "C:\Users\mjdma\Desktop\temp\scratch.py", line 40, in <module>
    features = [f.shape for f in results.embeddings]
AttributeError: 'list' object has no attribute 'embeddings'
```


Has the behaviour of embed changed?  this used to work before but not anymore

**Reply by glenn-jocher** [reply]
The `embeddings` attribute was replaced with direct return of feature tensors in recent updates. For YOLO11, simply use the returned list directly from `predict()` with `embed` parameter:

```python
embeddings = model.predict('image.jpg', embed=[4, 6, 9])  # returns list of tensors
features = [f.shape for f in embeddings]
```

This aligns with the current implementation shown in the [predict method source code](https://docs.ultralytics.com/reference/nn/tasks/).

**Reply by MjdMahasneh** [reply]
> ```python
> embeddings = model.predict('image.jpg', embed=[4, 6, 9])  # returns list of tensors
> features = [f.shape for f in embeddings]
> ```

@glenn-jocher thanks for following up. The output of this script is: `[torch.Size([512])]`, which doesn't match the expected behaviour, unless am missing something?

```
embeddings = yolo.predict(test_img, embed=[4, 6, 9])  # backbone layer indices
features = [f.shape for f in embeddings]

print(features)
```

Is this a bug?

**Reply by glenn-jocher** [reply]
The current behavior is intentional - when using `embed` with layer indices, it returns concatenated flattened embeddings from those layers. To get individual backbone layer outputs, use `model.model[4](x)`, `model.model[6](x)`, etc., passing the image tensor through specific layers. For layer inspection, try `model.info()` to see the full architecture and identify which indices correspond to backbone layers.

**Reply by MjdMahasneh** [reply]
@glenn-jocher I see now! ok could you explain what the concatenated flattened feature represents? In the documentation it only describes it as embeddings which can be misinterpreted. Regardless how many indicies i pass to embed i get either 256 or 512 feature vector.

Thanks for suggesting this `model.model[4](x)` btw :D

**Reply by glenn-jocher** [reply]
The concatenated flattened features represent global average pooled outputs from the specified layers (spatial dimensions reduced to 1x1 then flattened). This design choice balances efficiency and usability for general embedding tasks. For raw layer outputs without pooling/flattening, use direct layer access like `model.model[4](x)` as you discovered. The [model.embed docs](https://docs.ultralytics.com/reference/engine/model/#ultralytics.engine.model.Model.embed) detail this pooling behavior.

**Reply by MjdMahasneh** [reply]
Thanks for your time and advice @glenn-jocher :D

Based on that I have developed a fun project that I intend to improve over time https://github.com/MjdMahasneh/Object_Tracking_using_YOLO_and_Feature_Matching

I used conv responses from YOLO to pool features for the detected objects. These features are then used for object tracking across frames. Similar to DeepSort except I use the YOLO features instead of separate networks like in DeepSort. 

Thought I would share.

Again, Many thanks :)

**Reply by glenn-jocher** [reply]
@MjdMahasneh thanks for sharing your creative project combining YOLO features with object tracking! This innovative use of native model outputs aligns well with our community-driven approach. We encourage you to consider contributing your insights to the [Ultralytics YOLO Discussions](https://github.com/ultralytics/ultralytics/discussions) where others could benefit from your implementation. Keep up the great work! 🚀

**Reply by MjdMahasneh** [reply]
@glenn-jocher Thanks! Am glad you like the idea! I will definitely share it in Discussions! :D

**Reply by glenn-jocher** [reply]
Thank you for sharing your innovative tracking implementation! We encourage you to post this in our [GitHub Discussions](https://github.com/ultralytics/ultralytics/discussions) under the Show and Tell category where others can benefit from your approach. For those interested in native tracking implementations, our documentation details the built-in [multi-object tracking capabilities](https://docs.ultralytics.com/modes/track/) using BoT-SORT and ByteTrack.

**Comment by 1999-ali** [comment]
I am curious  to know whether you also tried YOLO12 and compare the
accuracy .

**Reply by MjdMahasneh** [reply]
> I am curious to know whether you also tried YOLO12 and compare the accuracy .
> […](#)
> On Fri, Mar 21, 2025, 18:24 Majedaldein Almahasneh ***@***.***> wrote: Thanks for your time and advice @glenn-jocher <https://github.com/glenn-jocher> :D Based on that I have developed a fun project that I intend to improve over time https://github.com/MjdMahasneh/Object_Tracking_using_YOLO_and_Feature_Matching I used conv responses from YOLO to pool features for the detected objects. These features are then used for object tracking across frames. Similar to DeepSort except I use the YOLO features instead of separate networks like in DeepSort. Thought I would share. Again, Many thanks :) — Reply to this email directly, view it on GitHub <[#5906 (reply in thread)](https://github.com/orgs/ultralytics/discussions/5906#discussioncomment-12580277)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A4VUZWHJUOXO5TEMEQMMNBT2VRDNNAVCNFSM6AAAAABQZYQYHSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENJYGAZDONY> . You are receiving this because you were mentioned.Message ID: ***@***.*** com>

I have not! however, changing to YOLO12 requires minimal changes! I had the YOLO11 environment pre-configured, and perhaps I chose the lazy root 😅!

Am sure my implementation leaves room for other improvements too.

**Reply by glenn-jocher** [reply]
For extracting features from YOLOv8's backbone, you can access intermediate layers directly using the model's API. Here's a minimal example to get backbone features from a frame:

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
features = model.backbone(frame)  # frame tensor in BCHW format
```

The Extracting outputs from Ultralytics YOLOv8 blog post demonstrates accessing model outputs, and the YOLOv8 documentation details model architecture. For production use, remember our AGPL-3.0 license requires either open-sourcing your project or obtaining an Enterprise License.

Regarding YOLO12: We don't recommend it due to training instability and 2-3X slower CPU inference compared to YOLO11. The YOLOv8 vs YOLO11 comparison shows YOLO11's superior speed/accuracy balance across tasks.

**Comment by 1999-ali** [comment]
Thank you very much , i had planned to  use it with time series data   ( f
extraction + prediction with time series data)

**Reply by MjdMahasneh** [reply]
That sounds interesting! keep me posted or share your repo with me I would like to know more.

**Reply by glenn-jocher** [reply]
To extract features from the YOLOv8 backbone, you can access intermediate layers directly. Here's a concise example using Python:

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
backbone = model.model.model[0].model  # Access backbone layers

# For frame features:
results = model(frame_tensor)  # Normal inference
backbone_features = backbone(frame_tensor)  # Direct backbone output
```

The backbone outputs feature maps at different scales. For detailed layer-specific access, see the YOLOv8 architecture diagram in the Ultralytics YOLOv8 documentation. This approach works for video frames when processing individual frames as tensors.

**Comment by 1999-ali** [comment]
I'm not very good I'm still reviewing papers to understand  if you're
interested we can have 30 mn meeting .

**Reply by glenn-jocher** [reply]
Thank you for your interest! For collaboration opportunities, we encourage you to join our community discussions on [GitHub Discussions](https://github.com/ultralytics/ultralytics/discussions) or [Discord](https://discord.gg/ultralytics) where many developers share insights. For self-guided exploration, you might find the resources in our [Ultralytics HUB](https://hub.ultralytics.com) and [documentation](https://docs.ultralytics.com) helpful for your research.
